### PR TITLE
fix(core): Configureautoscaling calling autoscaling connections securitygroups in Cluster (under feature flag)

### LINKED ISSUE #22764: aws-ecs: Cluster.addAsgCapacityProvider doesn't add security groups to Cluster.connections like addAutoScalingGroup did
### Describe the issue

The docs for [cluster.addAutoScalingGroup](https://docs.aws.amazon.com/cdk/api/v1/docs/@aws-cdk_aws-ecs.Cluster.html#addwbrautowbrscalingwbrgroupautoscalinggroup-optionsspan-classapi-icon-api-icon-deprecated-titlethis-api-element-is-deprecated-its-use-is-not-recommended%EF%B8%8Fspan) point towards using [cluster.addAsgCapacityProvider](https://docs.aws.amazon.com/cdk/api/v1/docs/@aws-cdk_aws-ecs.Cluster.html#addwbrasgwbrcapacitywbrproviderprovider-options) instead, as the former is deprecated, and gone in v2.

They are nearly direct replacements.

The obvious - but wrong - replacement I started with was as follows:
Before:
```typescript
myCluster.addAutoScalingGroup(myAsg);
```
After:
```typescript
const myCapacityProvider = new ecs.AsgCapacityProvider(this, 'MyCapacityProvider', {
    autoScalingGroup: myAsg,
});
myCluster.addAsgCapacityProvider(myCapacityProvider);
```

However, this resulted in surprise issues with security groups elsewhere in my code (some code assumed an ECS Service had at least one security group in `.connections.securityGroups`, but it had zero, so errored.).

Turns out this is because `cluster.addAutoScalingGroup` adds security groups from the ASG's connections into the ECS Cluster's connections, but `cluster.addAsgCapacityProvider` does not.

See https://github.com/aws/aws-cdk/blob/c4e07a65f9580d3e39d4077d872bee6ad17d8924/packages/%40aws-cdk/aws-ecs/lib/cluster.ts#L385

So, in my example above, I needed to add:
```typescript
myCluster.connections.connections.addSecurityGroup(...myAsg.connections.securityGroups)
```

I don't really understand why this isn't done in `configureAutoScalingGroup` like all the other common setup is. I think it would be fine to add, as it won't break for someone who has already added the workaround like I have, as it seems that re-adding security groups to connections no-op (they don't get duplicated).

---

So I think one (or more) of the following could resolve this:
- the docs pointing from `addAutoScalingGroup` to `addAsgCapacityProvider` should highlight the ways in which they are different, and guide towards how to do a direct replacement.
- the options that can be passed to `addAsgCapacityProvider` should allow setting a `addSecurityGroupsFromAsgConnectionsToClusterConnections` or something, and when set to `true`, this does add security groups to the clusters connections.
- `Cluster.configureAutoScalingGroup` should call `this.connections.connections.addSecurityGroup(...autoScalingGroup.connections.securityGroups);`

---

Other related issues and PRs:
- https://github.com/aws/aws-cdk/pull/15107 this appears to have been an attempt to fix the issue, but the PR stalled
    - and the linked issue https://github.com/aws/aws-cdk/issues/14764
- https://github.com/aws/aws-cdk/issues/16260#issuecomment-937918710
- https://github.com/aws/aws-cdk/issues/15733

### Links

addAutoScalingGroup: https://docs.aws.amazon.com/cdk/api/v1/docs/@aws-cdk_aws-ecs.Cluster.html#addwbrautowbrscalingwbrgroupautoscalinggroup-optionsspan-classapi-icon-api-icon-deprecated-titlethis-api-element-is-deprecated-its-use-is-not-recommended%EF%B8%8Fspan

cluster.addAsgCapacityProvider: https://docs.aws.amazon.com/cdk/api/v1/docs/@aws-cdk_aws-ecs.Cluster.html#addwbrasgwbrcapacitywbrproviderprovider-options

### PR DESCRIPTION
**Pull request by homakk**
Configureautoscaling calling autoscaling connections securitygroups in Cluster (under feature flag)
image
>
> Describe the reason for this change, what the solution is, and any
> important design decisions you made. 
>
> Remember to follow the [CONTRIBUTING GUIDE] and [DESIGN GUIDELINES] for any
> code you submit.
>
> [CONTRIBUTING GUIDE]: https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md
> [DESIGN GUIDELINES]: https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md

Closes #22764
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/cx-api/FEATURE_FLAGS.md [hunk 1]
```diff
@@ -45,7 +45,7 @@ Flags come in three types:
 | [@aws-cdk/customresources:installLatestAwsSdkDefault](#aws-cdkcustomresourcesinstalllatestawssdkdefault) | Whether to install the latest SDK by default in AwsCustomResource | 2.60.0 | (default) |
 | [@aws-cdk/aws-codedeploy:removeAlarmsFromDeploymentGroup](#aws-cdkaws-codedeployremovealarmsfromdeploymentgroup) | Remove CloudWatch alarms from deployment group | 2.65.0 | (fix) |
 | [@aws-cdk/aws-rds:databaseProxyUniqueResourceName](#aws-cdkaws-rdsdatabaseproxyuniqueresourcename) | Use unique resource name for Database Proxy | 2.65.0 | (fix) |
-| [@aws-cdk/aws-apigateway:authorizerChangeDeploymentLogicalId](#aws-cdkaws-apigatewayauthorizerchangedeploymentlogicalid) | Include authorizer configuration in the calculation of the API deployment logical ID. | 2.66.0 | (fix) |
+| [@aws-cdk/aws-apigateway:authorizerChangeDeploymentLogicalId](#aws-cdkaws-apigatewayauthorizerchangedeploymentlogicalid) | Include authorizer configuration in the calculation of the API deployment logical ID. | V2NEXT | (fix) |
```
Comment (bvtujo) [inline packages/@aws-cdk/cx-api/FEATURE_FLAGS.md:48]: Can you make sure that you don't make any changes to the feature flags file? This doesn't seem to be necessary for this fix.

**Review by bvtujo — COMMENTED** [review]
Looks like the integ tests in the aws-ecs-patterns package are failing as a result of this change. @TheRealAmazonKendra this is a p1 fix but it may be breaking for other customers--is there any guidance on whether we need to add a feature flag here?

File: packages/@aws-cdk/cx-api/FEATURE_FLAGS.md (hunk 1 above)
Comment (homakk) [reply packages/@aws-cdk/cx-api/FEATURE_FLAGS.md]: Ok. I revert this changes back and pushed those changes

**Comment by TheRealAmazonKendra** [comment]
> #22764

Yes, I believe this will need a feature flag. Also, this is a fix, not a chore.

File: packages/@aws-cdk/cx-api/lib/features.ts [hunk 2]
```diff
@@ -236,6 +237,23 @@ export const FLAGS: Record<string, FlagInfo> = {
...
+    type: FlagType.ApiDefault,
+    summary: 'Add security group through "configureAutoScalingGroup"',
+    detailsMd: `
+    ConfigureAutoScalingGroup, currently determine for the ecs security group  of
+    a CfnService if a security group is not provided. The result of this is that on every
+    deployment, the security group is reset to the fixed, even if it was autoscaled.
+    
+    If this flag is not set, the cluster.addAutoScalingGroup adds security groups from 
+    the ASG's connections into the ECS Cluster's connections, but cluster.addAsgCapacityProvider does not `,
+    introducedIn: { v1: '1.92.0' },
```
Comment (bvtujo) [inline packages/@aws-cdk/cx-api/lib/features.ts:251]: Please read the feature flag guidelines [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#feature-flags) and ensure this PR meets them. For example, you need to be careful that this `introducedIn` field is correct. It should be `{ v2: 'V2NEXT' }`.

Also, make sure the defaults are right. Basically please be sure that this is exactly correct, it's very important to be precise here.

File: packages/@aws-cdk/cx-api/lib/features.ts [hunk 3]
```diff
@@ -236,6 +237,23 @@ export const FLAGS: Record<string, FlagInfo> = {
     compatibilityWithOldBehaviorMd: 'You can pass `desiredCount: 1` explicitly, but you should never need this.',
   },
 
+  //////////////////////////////////////////////////////////////////////
+  [ECS_ADD_SECURITY_GROUP]: {
```
Comment (bvtujo) [inline packages/@aws-cdk/cx-api/lib/features.ts:241]:
```suggestion
  [ECS_ADD_SECURITY_GROUPS_TO_ASG_CAPACITY_PROVIDERS]: {
```
let's call this feature flag something more descriptive

File: packages/@aws-cdk/cx-api/lib/features.ts [hunk 4]
```diff
@@ -53,6 +53,7 @@ export const SECRETS_MANAGER_PARSE_OWNED_SECRET_NAME = '@aws-cdk/aws-secretsmana
 export const KMS_DEFAULT_KEY_POLICIES = '@aws-cdk/aws-kms:defaultKeyPolicies';
 export const S3_GRANT_WRITE_WITHOUT_ACL = '@aws-cdk/aws-s3:grantWriteWithoutAcl';
 export const ECS_REMOVE_DEFAULT_DESIRED_COUNT = '@aws-cdk/aws-ecs-patterns:removeDefaultDesiredCount';
+export const ECS_ADD_SECURITY_GROUP = '@aws-cdk/aws-ecs-patterns:addSecurityGroup';
```
Comment (bvtujo) [inline packages/@aws-cdk/cx-api/lib/features.ts:56]: Let's call this something more descriptive. See below.

File: packages/@aws-cdk/cx-api/lib/features.ts [hunk 5]
```diff
@@ -236,6 +237,23 @@ export const FLAGS: Record<string, FlagInfo> = {
...
+  [ECS_ADD_SECURITY_GROUP]: {
+    type: FlagType.ApiDefault,
+    summary: 'Add security group through "configureAutoScalingGroup"',
+    detailsMd: `
+    ConfigureAutoScalingGroup, currently determine for the ecs security group  of
+    a CfnService if a security group is not provided. The result of this is that on every
+    deployment, the security group is reset to the fixed, even if it was autoscaled.
+    
+    If this flag is not set, the cluster.addAutoScalingGroup adds security groups from 
+    the ASG's connections into the ECS Cluster's connections, but cluster.addAsgCapacityProvider does not `,
```
Comment (bvtujo) [inline packages/@aws-cdk/cx-api/lib/features.ts:250]:
```suggestion
    ConfigureAutoScalingGroup currently does not connect the ASG security group to the cluster's security group. The result of this is that on new deployments, EC2 instances which have been autoscaled can have their security groups reconfigured and lose connectivity to the ECS cluster. This feature flag enables the correct behavior. 
    
    If this flag is not set,  `cluster.addAsgCapacityProvider()` does not correctly configure the autoscaling group's Security Groups to work with the ECS cluster. If the flag is set, the ASG is correctly configured.`,
```

File: packages/@aws-cdk/cx-api/lib/features.ts (hunk 4 above)
Comment (homakk) [reply packages/@aws-cdk/cx-api/lib/features.ts]: changed to  ECS_ADD_SECURITY_GROPU_TO_ASG_CAPACITY_PROVIDERS

File: packages/@aws-cdk/cx-api/lib/features.ts (hunk 3 above)
Comment (homakk) [reply packages/@aws-cdk/cx-api/lib/features.ts]: Got it and changed .... you can see in the latest commit

File: packages/@aws-cdk/cx-api/lib/features.ts (hunk 2 above)
Comment (homakk) [reply packages/@aws-cdk/cx-api/lib/features.ts]: changed 
```
introducedIn: { v2: 'V2NEXT' },
    defaults: { v2: true },
    recommendedValue: true,
```

**Comment by Naumel** [comment]
>AWS CodeBuild CI Report
>Result: FAILED

```
FAIL test/resource-pool.test.js
@aws-cdk-testing/cli-integ:   â— double dispose throws
@aws-cdk-testing/cli-integ:     expect(received).toThrow()
@aws-cdk-testing/cli-integ:     Received function did not throw
@aws-cdk-testing/cli-integ:       33 |
@aws-cdk-testing/cli-integ:       34 |   await lease.dispose();
@aws-cdk-testing/cli-integ:     > 35 |   expect(() => lease.dispose()).toThrow();
@aws-cdk-testing/cli-integ:          |                                 ^
@aws-cdk-testing/cli-integ:       36 | });
@aws-cdk-testing/cli-integ:       37 |
@aws-cdk-testing/cli-integ:       38 | test('somewhat balance', async () => {
@aws-cdk-testing/cli-integ:       at Object.<anonymous> (test/resource-pool.test.ts:35:33)
@aws-cdk-testing/cli-integ:   â— somewhat balance
@aws-cdk-testing/cli-integ:     Calling dispose() on an already-disposed lease.
@aws-cdk-testing/cli-integ:       93 |       dispose: async () => {
@aws-cdk-testing/cli-integ:       94 |         if (disposed) {
@aws-cdk-testing/cli-integ:     > 95 |           throw new Error('Calling dispose() on an already-disposed lease.');
@aws-cdk-testing/cli-integ:          |                 ^
@aws-cdk-testing/cli-integ:       96 |         }
@aws-cdk-testing/cli-integ:       97 |         disposed = true;
@aws-cdk-testing/cli-integ:       98 |         return this.returnValue(value).finally(() => process.stderr.write(`Lease returned by ${process.pid}: ${value}`));
@aws-cdk-testing/cli-integ:       at Object.dispose (lib/resource-pool.ts:95:17)
@aws-cdk-testing/cli-integ:       at test/resource-pool.test.ts:35:22
@aws-cdk-testing/cli-integ:       at Object.<anonymous> (../../../node_modules/expect/build/toThrowMatchers.js:83:11)
@aws-cdk-testing/cli-integ:       at Object.toThrow (../../../node_modules/expect/build/index.js:382:21)
@aws-cdk-testing/cli-integ:       at Object.<anonymous> (test/resource-pool.test.ts:35:33)
```

**Comment by TheRealAmazonKendra** [comment]
@Mergifyio update

**Review by bvtujo — APPROVED** [review]

**Pull request closed without merging**